eng/packages/http-client-csharp-mgmt/emitter/src/resource-detection.ts

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+import {
+  CodeModel,
+  InputClient,
+  InputModelType
+} from "@typespec/http-client-csharp";
+import {
+  calculateResourceTypeFromPath,
+  ResourceMetadata
+} from "./resource-metadata.js";
+import { DecoratorInfo } from "@azure-tools/typespec-client-generator-core";
+
+// https://github.com/microsoft/typespec/blob/main/packages/rest/README.md#parentresource
+export const parentResource = "TypeSpec.Rest.@parentResource";
+export const parentResourceRegex = "TypeSpec\\.Rest\\.@parentResource";
+
+// https://github.com/Azure/typespec-azure/blob/main/packages/typespec-azure-resource-manager/README.md#armresourceoperations
+export const armResourceOperations =
+  "Azure.ResourceManager.@armResourceOperations";
+export const armResourceOperationsRegex =
+  "Azure\\.ResourceManager\\.@armResourceOperations";
+
+// https://github.com/Azure/typespec-azure/blob/main/packages/typespec-azure-resource-manager/README.md#armResourceRead
+export const armResourceRead = "Azure.ResourceManager.@armResourceRead";
+export const armResourceReadRegex =
+  "Azure\\.ResourceManager\\.@armResourceRead";
+
+export const armResourceCreateOrUpdate =
+  "Azure.ResourceManager.@armResourceCreateOrUpdate";
+export const armResourceCreateOrUpdateRegex =
+  "Azure\\.ResourceManager\\.@armResourceCreateOrUpdate";
+// https://github.com/Azure/typespec-azure/blob/main/packages/typespec-azure-resource-manager/README.md#singleton
+export const singleton = "Azure.ResourceManager.@singleton";
+export const singletonRegex = "Azure\\.ResourceManager\\.@singleton";
+
+// TODO: add this decorator to TCGC
+export const resourceMetadata = "Azure.ClientGenerator.Core.@resourceSchema";
+export const resourceMetadataRegex =
+  "Azure\\.ClientGenerator\\.Core\\.@resourceSchema";
+
+export function updateClients(codeModel: CodeModel) {
+  // first we flatten all possible clients in the code model
+  const clients = getAllClients(codeModel);
+
+  // first pass to update all the clients with their own information
+  const metadata: Map<InputClient, ResourceMetadata> = new Map();
+  for (const client of clients) {
+    gatherResourceMetadata(client, metadata);
+  }
+
+  // add the decorator to the client
+  for (const client of clients) {
+    const resourceMetadata = metadata.get(client);
+    if (resourceMetadata) {
+      addResourceMetadata(client, resourceMetadata);
```

Comment:
We need to iterate the clients again because we need to get the resource hierarchy(i.e., ParentResource) in the first loop.
Otherwise, we can just add the metadata in the first loop.
Can we add some comments for this?

Reply:
I updated on the above function as a process description